pkg/tuned/controller.go

```diff
@@ -1046,23 +1271,116 @@ func (c *Controller) updateTunedProfile() (err error) {
 		}
 	}
 
+	return c.updateTunedProfileStatus(context.TODO(), change)
+}
+
+func (c *Controller) updateTunedProfileStatus(ctx context.Context, change Change) error {
+	activeProfile, err := getActiveProfile()
+	if err != nil {
+		return err
+	}
+
+	profile, err := c.listers.TunedProfiles.Get(c.nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to get Profile %s: %v", c.nodeName, err)
+	}
+
+	var message string
+	wantsDeferred := util.HasDeferredUpdateAnnotation(profile.Annotations)
+	isApplied := (c.daemon.profileFingerprintUnpacked == c.daemon.profileFingerprintEffective)
+	daemonStatus := c.daemon.status
+
+	klog.V(4).Infof("daemonStatus(): change: deferred=%v applied=%v nodeRestart=%v", wantsDeferred, isApplied, change.nodeRestart)
+	if (wantsDeferred && !isApplied) && !change.nodeRestart { // avoid setting the flag on updates deferred -> immediate
+		daemonStatus |= scDeferred
+		recommendProfile, err := TunedRecommendFileRead()
+		if err == nil {
+			klog.V(2).Infof("updateTunedProfileStatus(): recommended profile %q (deferred)", recommendProfile)
+			message = c.daemonMessage(change, recommendProfile)
+		} else {
+			// just log and carry on, we will use this info to clarify status conditions
+			klog.Errorf("%s", err.Error())
+		}
+	}
+
+	statusConditions := computeStatusConditions(daemonStatus, message, profile.Status.Conditions)
+	klog.V(4).Infof("computed status conditions: %#v", statusConditions)
+	c.daemon.status = daemonStatus
+
 	if profile.Status.TunedProfile == activeProfile &&
 		conditionsEqual(profile.Status.Conditions, statusConditions) {
-		klog.V(2).Infof("updateTunedProfile(): no need to update status of Profile %s", profile.Name)
+		klog.V(2).Infof("updateTunedProfileStatus(): no need to update status of Profile %s", profile.Name)
 		return nil
 	}
 
 	profile = profile.DeepCopy() // never update the objects from cache
 
 	profile.Status.TunedProfile = activeProfile
 	profile.Status.Conditions = statusConditions
-	_, err = c.clients.Tuned.TunedV1().Profiles(operandNamespace).UpdateStatus(context.TODO(), profile, metav1.UpdateOptions{})
+	_, err = c.clients.Tuned.TunedV1().Profiles(operandNamespace).UpdateStatus(ctx, profile, metav1.UpdateOptions{})
 	if err != nil {
 		return fmt.Errorf("failed to update Profile %s status: %v", profile.Name, err)
 	}
 	return nil
 }
 
+func (c *Controller) storeDeferredUpdate(deferredFP string) (derr error) {
```

Comment:
I believe this deserves some comments and rationale for using temporary files.  Presumably trying to be as atomic as possible not to end up with a corrupt persistent deffered update file?

Reply:
done